lib/learn_open/environments/ide_environment.rb

```diff
@@ -2,43 +2,55 @@ module LearnOpen
   module Environments
     class IDEEnvironment < BaseEnvironment
       def open_readme(lesson)
-        if valid?(lesson)
+        when_valid(lesson) do
+          warn_if_necessary(lesson)
           io.puts "Opening readme..."
           run_custom_command(:browser_open, {url: lesson.to_url})
-        else
-          run_custom_command(:open_lab, {lab_name: lesson.name})
         end
       end
 
       def open_jupyter_lab(lesson, location, editor)
-        if valid?(lesson)
+        when_valid(lesson) do
+          warn_if_necessary(lesson)
           io.puts "Opening Jupyter Lesson..."
           run_custom_command(:browser_open, {url: lesson.to_url})
-        else
-          run_custom_command(:open_lab, {lab_name: lesson.name})
         end
       end
 
       def open_lab(lesson, location, editor)
-        case lesson
-        when LearnOpen::Lessons::IosLesson
-          super
-        when -> (lesson) { valid?(lesson) }
-          download_lesson(lesson, location)
-          open_editor(lesson, location, editor)
-          start_file_backup(lesson, location)
-          install_dependencies(lesson, location)
-          notify_of_completion
-          open_shell
+        when_valid(lesson) do
+          warn_if_necessary(lesson)
+          case lesson
+          when LearnOpen::Lessons::IosLesson
+            super
+          when -> (l) { valid?(l) }
```

Comment:
Ah, I still think this proc/lambda in a case statement is crazy, but here is a [post](http://blog.honeybadger.io/rubys-case-statement-advanced-techniques/) for anyone who is curious about why/how this works.